File: 547.py
class Solution(object):
    def findCircleNum(self, isConnected):
        n = len(isConnected)
        count = 0
        visited = bytearray(n)
        
        for i in range(n):
            if not visited[i]:
                count += 1
                visited[i] = 1
                queue = [i]
                while queue:
                    city = queue.pop()
                    for neighbor in range(n):
                        if isConnected[city][neighbor] and not visited[neighbor]:
                            queue.append(neighbor)
                            visited[neighbor] = 1
        return count
    
"""
isConnected = [[1, 1, 0], [1, 1, 0], [0, 0, 1]]
s = Solution()
res = s.findCircleNum(isConnected)
print(res)
"""

File: src/547.cpp
#include <vector>

class Solution {
public:

    void dfs(std::vector<std::vector<int>>& isConnected, std::vector<bool>& visited, int i){
        visited[i] = true;
        for(int j = 0; j < isConnected.size() ; ++j){
            if(isConnected[i][j] == 1 and !visited[j]){
                dfs(isConnected, visited, j);
            }
        }
    }
    int findCircleNum(std::vector<std::vector<int>>& isConnected) {
        int n = isConnected.size();
        std::vector<bool> visited(n, false);
        int provinces = 0;

        for(int i=0; i < n; ++i){
            if(!visited[i]){
                dfs(isConnected, visited, i);
                ++provinces;
            }
        }
        return provinces;
    }
};
